[KEP] storage api version migration

caesarxuchao opened the pull request:
This KEP proposes a tool that migrates the stored data in etcd to the latest storage version. The KEP is based on the [doc](https://docs.google.com/document/d/16HWEjqnlEc2PJnS4D7TVbHoNyVlyWpjx0zRsFNyxDsI/edit#heading=h.fap52hhsdgdt) which was discussed in the Aug. 1, 2018 sig apimachinery meeting. The KEP added more details on the tool's workflow and development timeline.

/sig api-machinery
/assign @lavalamp @deads2k

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+
+## Summary
+
+We propose to extend and improve the [oc adm migrate stroage][] command line
```
Done.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+### Goals
+
+A successful storage version migration tool must:
+* only cause insignificant load to apiservers. For example, if the master has
```
Got it. The `oc adm migrate` already has rate limiting based on bandwidth, though I don't know whether bandwidth or qps is a better metric.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+As to the deliverables,
+* in the short term, providing system administrators with a tool to migrate
+  the API objects to the proper storage versions. 
+* in the long term, automating the migration without further burdening
```
Done. Explicitly say the final goal is automating the migration for built-in APIs, CRD, and aggregated apis. But at alpha, CRD is ignored.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+cluster. "kube-migrator" uses the same kubeconfig file as kubectl does.
+
+The "kube-migrator controller" does the following:
+* creates a `progress` custom resource (CR).
```
Added `in the migration.k8s.io group`.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+* uses the discovery client to discover the supported APIs.
+* lists all CRDs (custom resource definitions).
+* stores the discovered resource types, excluding resources defined by CRDs, in
+  the "progress" CR.
```
In the alpha stage, the migration tool will be manually invoked, so I just say `the admin should invoke the tool after all the apiservers are at the desired version, including both kube apiservers and aggregated apiservers`

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
````diff
@@ -0,0 +1,244 @@
+}
+```
+
+The "migration worker" runs `oc adm migrate storage --include=<resource type>`
````
Changed to "runs the equivalence of `oc adm migrate storage --include=<resource type>`". I keep the old name to show how the existing oc tool fits in.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+
+Kubernetes replicaset controller restarts the "migration controller" pod
+if it fails. When the "migration controller" starts,
+* it checks if there exists any `progress` object that matches the current
```
The migrator for each resource type stores continue tokens. The `progress` object in addition needs to record if all resource types have been migrated.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+native resources, the migrator only needs to run when the apiservers upgrade to
+a different minor version. Thus, the migration can be triggered automatically
+when the version stored in the `progress` object differs from what it returned
+at kube apiserver's `/version` endpoint. 
```
Removed apiserver version from the alpha plan, and discussed in the automation section.

I don't think an incrementing counter is necessary, the command line tool just deletes old deployments before creating a new one. See the updated "workflow" section.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+HA apiservers agreement problem is a precondition for the migration automation.
+
+For CRD, because the CRD exposes all versions that are ever persisted in the
+storage since last migration, the migration should migrate any CRD that has
```
Changed the entire section. Explicitly state that in alpha we don't do automation, and we don't handle CRD.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+storage since last migration, the migration should migrate any CRD that has
+more than two persisted storage versions. 
+
+For aggregated apiservers, they can expose their version to trigger the tool
```
Rewrote the section.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+If the rate limiting is not tuned well, the migration can overload the
+apiserver. Users can delete the migration controller and the migration
+jobs to mitigate.
+
```
Added a paragraph suggesting rollback and redo the migration if that happens.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+
+## Open questions
+
+* Resources can migrate across API groups. For example, `deployment` was
```
I'll expand in the next iteration.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+“active management”) is another propose. "Active management" is different to
+this proposal ("kube-migrator") in that it requires the apiservers to expose the
+storage versions. The advantage is that the "active management" controller only
+has to touch the resource types whose storage versions have changed, thus more
```
Got it.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,244 @@
+### update-storage-objects.sh
+Nevertheless, the Kubernetes repo has a update-storage-objects.sh. It is not
+adopted by the community because it is not production ready: no rate limiting,
+hard-coded resource types, no persisted migration states.
```
Done.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,291 @@
+jobs to mitigate.
+
+Before upgrading or downgrading the cluster, the cluster administrator MUST
+check the "progress" object, and make sure every resource type is marked as
```
We can do slightly better than this. It can state separately whether roll-back and roll-forward is safe.

E.g., if you roll-forward, and then roll-back before the tool is finished, that is actually fine. You just can't roll-forward twice in a row without waiting, nor can you roll back twice in a row.

The tool has to track state to understand whether the operations in progress are rolling forward or backwards to support this.

I think it might actually be quite important, especially in large clusters, to be able to do an immediate rollback without waiting for a data migration.

liggitt replied:
> can we deprecate --storage-versions? It's defined in pkg/kubeapiserver. Kuberentes is following the 4-step deprecation policy, I don't see in what case the --storage-versions flag is useful

I agree with this. It seems like flexibility we don't need, that actually harms our ability to reason about what is safe to do as we maintain kube-apiserver in future releases.

caesarxuchao replied:
@lavalamp @deads2k what do you think about removing ` --storage-versions`? I opened https://github.com/kubernetes/kubernetes/pull/67678 to remove the flag, but it's more complicated than I expected, can you let me know early if you think it's a bad idea?

lavalamp replied:
removing `--storage-versions` means everyone has to use the default. How would you e.g. do a safe downgrade if someone skipped a step in the version upgrade process?

I think most people should use the default.

We can't remove the flag without a deprecation process unless we exhaustively demonstrate that no one is using it.

caesarxuchao replied:
> How would you e.g. do a safe downgrade if someone skipped a step in the version upgrade process?

Not sure if I follow, could you be more specific?

> We can't remove the flag without a deprecation process unless we exhaustively demonstrate that no one is using it.

How about keeping a noop `--storage-versions` flag for one more release?

side note: if users are using this flag, they need to manually inspect this flag every time they upgrade anyway, because the API version in the flag value can be deprecated.

lavalamp replied:
One more release is not our deprecation policy. Without looking, I think we need at least two, maybe more.

If someone introduces a v2 and also accidentally sets it as the storage version in the same release, it is impossible to downgrade safely without use of the storage version flag.

caesarxuchao replied:
> If someone introduces a v2 and also accidentally sets it as the storage version in the same release, it is impossible to downgrade safely without use of the storage version flag.

We should write tests to prevent this kind of bugs. However, as we don't expose the storage versions, I don't know how to write such tests.

The deprecation [policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli) says 6 months.

caesarxuchao replied:
Just realized that after we remove the `--storage-versions`, a group's preferred version showed in the discovery API equals the storage version, then we don't need to argue if we should expose the storage versions.. Is there any other reason why storage version might diverge from preferred version?

lavalamp commented on sig-api-machinery/README.md:
```diff
@@ -70,6 +70,9 @@ The following subprojects are owned by sig-api-machinery:
     - https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/master/OWNERS
     - https://raw.githubusercontent.com/kubernetes-sigs/controller-tools/master/OWNERS
     - https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/OWNERS
+- **server-upgrade-downgrade**
```
I think this is (going to be) a control plane feature--there's a section for that on line 42.

deads2k replied:
> I think this is (going to be) a control plane feature--there's a section for that on line 42.

I must need to re-read.  I remember talking about creating this as a separate tool.  It allowed evolution outside of kube releases and was a clear indication that it was indeed separate from core.

deads2k replied:
> I think this is (going to be) a control plane feature--there's a section for that on line 42.
> 
> I must need to re-read. I remember talking about creating this as a separate tool. It allowed evolution outside of kube releases and was a clear indication that it was indeed separate from core.

Or do you just mean, "yes it will be a separate repo, but it falls into an existing bucket and we can link the repo in the existing section"?

lavalamp replied:
> it will be a separate repo, but it falls into an existing bucket and we can link the repo in the existing section

Exactly. It will be a control plane feature when it grows up, no matter how it is invoked or where the code lives.

deads2k commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,291 @@
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  // A list of all the resources as suggested by the apiservers
+  Resources []Resource
```
The multiplicity here seems odd.  I would expect to have one instance per resource with both spec and status.

Also, shall we split the API and design for the controller out of this first pass?  I accept that such a controller can be built, but I don't see a reason to block the creation of the command line tool on the shape of the controller to come.

deads2k replied:
I think the API should be drivable by an external actor.  I should be able to indicate, "migrate storage for this resource".  That would bump a generation and then the actual migration would reflect a status generation to indicate that migration was complete.  Given an externally driveable change, I suspect you'd also want to be able indicate why you bumped it.

The granularity of both desired and actual state is at the level of resource.  Combining them makes a singleton resource which is generally odd, makes conflicts more likely, and generally means "bad" updates are more likely.

lavalamp replied:
I agree with @deads2k about no singleton resources. The "drivable by an external actor" part implies further creation of such an actor. It makes each part of the system more focused and hence composable, at the cost of more moving pieces. Overall I think it's probably slightly better to do it that way.

caesarxuchao commented:
Moved the repo creation request to https://github.com/kubernetes/community/pull/2568.

caesarxuchao commented:
@deads2k @lavalamp PTAL. Thanks.

Mainly changed the following:
* Addressed deads2k's comments on the singleton API. 
* Explicitly separated the alpha workflow (requiring admin intervention) from beta workflow (automated)
* Added the idea of expanding the discovery API with `StorageUID` to identify resources that share storage.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+### Goals
+
+A successful storage version migration tool must:
+* work for Kubernetes built-in APIs, custom resources (CR), and aggregated APIs.
```
I'm not sure if we should handle aggregated APIs, which can have their own storage conventions. The read-then-write operations might not migrate the stored data. The API can even be non-idempotent.

caesarxuchao replied:
The proposal isn't final yet. 

I want to discuss if storage migrator should take care of aggregated apis. If so, let's decide on the two alternatives listed in the `Beta workflow - Automation` section. Then I'll update the proposal.

caesarxuchao commented:
/assign @yliaog

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+### Risks and Mitigations
+
+The migration process does not change the objects, so it will not pollute
+existing data.
```
"Before upgrading or downgrading the cluster, the cluster administrator must run
the `kube-migrator progress` to make sure the migration has completed".

yliaog replied:
why not having this check automatically done inside kube-migrator? what if the admin forget to run "kube-migrator progress"?

caesarxuchao replied:
Perhaps I missed your point. "upgrading or downgrading the cluster" are done via cloud-specific tools. We cannot force those tools to run `kube-migrator progress`.

lavalamp replied:
We probably need the triggering mechanism to cancel an in-progress migration if the trigger changes (and then start a new one). It's e.g. safe for user to go forward and then back without waiting for the migration to complete.

But actually I don't know how to make this safe for a single object write (read on version X, but by the time there's a write it's version Y). You need something like a precondition on the api micro version for that to be 100% safe.

caesarxuchao replied:
> We probably need the triggering mechanism to cancel an in-progress migration if the trigger changes (and then start a new one). 

I think it's a general requirement to have a single migration job running for each resource type. I don't understand why it's a specific requirement of enabling fast roll back.

I didn't follow the second paragraph.

caesarxuchao replied:
Side note: to enable fast rollback, `migration.status` needs to reflect what is the currently stored in etcd, either the stored version, the hashed stored version, or the apiserver version (probably isn't safe in the future). I didn't put it in the proposal because we haven't reached consensus on the triggers.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+
+If the rate limiting is not tuned well, the migration can overload the
+apiserver. Users can delete the migration controller and the migration
+jobs to mitigate.
```
In the workflow added "makes sure there is only one **migration worker** `job` at a time, to avoid overloading the apiserver."

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+
+Before upgrading or downgrading the cluster, the cluster administrator MUST
+list the `migration` objects, and make sure every resource type is marked as
+completed. Otherwise the apiserver can crash. To mitigate, the cluster
```
Added an explanation, "because it cannot interpret the serialized data in etcd".

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+completed. Otherwise the apiserver can crash. To mitigate, the cluster
+administrator can rollback the apiserver to the old version, and wait for the
+migration to complete. If the apiserver does not crash, the `migration` will be
+corrupted, administrator needs to re-launch the **kube-migrator** to recover.
```
We are concerned that there is no one-to-one mapping between apiserver version and storage versions, because user can configure `--storage-versions` flag, though the flag will be removed in 1.13.
Another concern is that the aggregated apiserver doesn't have a server version. If we can agree on either forcing aggregated apiserver follows semver, or that the storage migration doesn't handle aggregated APIs by default, we can implement what you suggested.

caesarxuchao replied:
> it is not clear why 'migration' will be corrupted

Expanded. "If the apiserver does not crash after upgrading or downgrading, the `migration` objects lose track of migration progress, because the default storage versions might changed after upgrading".
or downgrading, but no one increments the `migration.spec.generation`.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+TODO: it is safe to rollback an apiserver to the previous version without
+waiting for the migration to complete. It is only unsafe to roll-forward or
+rollback twice. We need to design how to record the previous version, especially
+for the aggregated apiservers.
```
I think the blocker is aggregated apiservers. We need to either expose their versions, or decide not to take care of aggregate apiservers. Or we expose the storage versions of all resources. Basically the two alternatives in section `Beta workflow - Automation`.

Not sure what you mean by "guarded against failures in the migrator tool".

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+rollback twice. We need to design how to record the previous version, especially
+for the aggregated apiservers.
+
+## Graduation Criteria
```
Yeah, it's a roadmap. Keeping the title as "graduation criteria" because that' required of a KEP.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+* The migration needs to run every time the master is upgraded to a
+  different minor version, which happens once per quarter. 
+* **kube-migrator** ignore resources that only have only one version. If the
+  4-step deprecation [policy][] is followed, v1 and v2 only co-exist for 2
```
In the "view" mode I can click through the hyperlink. Can you double check?

yliaog replied:
I can view the link now.

where are the 4-steps? the deprecation policy doc lists a few rules, no 4-steps

caesarxuchao replied:
Removed this point. v1 and v2 can co-exist 4 quarters. The cost argument isn't critical after all. Safety is.

yliaog commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,360 @@
+
+On start-up, the **kube-migrator controller** does the following,
+* discovers all supported APIs.
+* lists all custom resource definitions, excludes them from the API list
```
so only CRDs are excluded? what about aggregated apis?

caesarxuchao replied:
Aggregated apis will be included in the first iteration, see line 90.

yliaog commented on keps/sig-api-machinery/storage-migration.md:
````diff
@@ -0,0 +1,346 @@
+```
+
+[APIResource]:https://github.com/kubernetes/kubernetes/blob/d38fb2c6451447ad0ecbdfed971a563419d60883/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L874
+
````
works now

yliaog commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+
+Kubernetes **replicaset controller** restarts the **migration controller** `pod`
+if it fails. When the **migration controller** starts, it resumes migration
+by reading the existing `migration` objects.
```
i'm still confused. Below, it lists all the tasks done at the "kube-migrator controller" start-up time. My question is how does it know it is the first time (started by kube-migrator command) vs a second time (it is killed and restarted by the replicaset controller)

==================================================
On start-up, the kube-migrator controller does the following,

discovers all supported APIs.
lists all custom resource definitions, excludes them from the API list retrieved in the previous step.
for each non-custom resource, creates a migration CR (see the API section for the schema).
for each migration instance, increments the migration.spec.generation, and logs the reason.

caesarxuchao replied:
Sorry for the confusion.

Updated this section to:
```
Kubernetes **replicaset controller** restarts the **migration controller** `pod`
if it fails. The restarted **migration controller** follows the same bootstrap
workflow as described in the [workflow section][]. Because the **migration
controller** adopts existing `migration` objects, it can resume from where
it left off.
```

and updated the workflow section to
```
On start-up, the **kube-migrator controller** does the following,
...
* for each non-custom resource, creates a `migration` CR (see the [API
  section][] for the schema) if it doesn't exist already.
* for each newly created `migration` instance, increments the
  `migration.spec.generation`, and logs the reason.
...
```

yliaog commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+of Kubernetes distributions.
+
+The automated migration should work for Kubernetes built-in resource types,
+custom resources, and aggregated resources.
```
ok, if custom resources are not that different from native resources, what prevents it from being included in the alpha phase?

caesarxuchao replied:
In the alpha phase, the migrator tool is manually invoked by cluster admins. They don't control the CRD lifecycle, so they don't know when to migrate the stored CRs.

In the beta iteration, where we want to automate the migrator invocation, we will build another component which monitors the CRD spec to figure out when their storage versions change, and then updates `migration.spec.generation`.

yliaog commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,360 @@
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#api
+
+The cluster admin can run the `kube-migrator progress` command to check if the
```
so kube-migrator has 'install' and 'progress' commands, what about 'uninstall' that can clean up everything related to the migration.

caesarxuchao replied:
Yeah, we can add such a subcommand, would be trivial. I will skip it in the KEP.

lavalamp replied:
Can this state be kept in a CRD, such that I can use kubectl to read the current progress? It seems contra our general direction to introduce new command line tools.

caesarxuchao replied:
Removed the command, user can just run `kubectl get migrations -o jsonpath="{.status.completed}"`.

caesarxuchao commented:
I have finished the iterations with @yliaog. 

We think the controversial part of this KEP is how to handle aggregated APIs, which can have their own storage conventions. The read-then-write operations might not migrate the stored aggregated APIs data. The API can even be non-idempotent. So at least we should make aggregated API's participation optional.

To handle aggregated APIs, we need to decide on the two alternatives in the "Beta workflow -  Automation" section.

Handling aggregated API isn't a blocker for the alpha phase. The tool will not distinguish aggregated APIs from built-in APIs in the alpha phase.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,360 @@
+* aggregated apiservers follow the same API deprecation [policy][].
+
+Then the migration of built-in resources and aggregated resources can be driven
+by the changes of apiservers versions.
```
I'm pretty against driving storage changes on apiserver binary version changes. Many times when it changes there's no storage change needed.

caesarxuchao replied:
> Many times when it changes there's no storage change needed.

Migrating objects whose storage version hasn't changed doesn't impact correctness. The extra workload is also tolerable.

If storage versions can change when apiserver binary versions do *not* change, that will impact correctness. Can you suggest cases where we need to do that?

deads2k replied:
> I'm pretty against driving storage changes on apiserver binary version changes. Many times when it changes there's no storage change needed.

Given the fairly low cost of migration and the fairly low frequency of binary versions, it seems reasonable to kick the storage migration globally when it happens.  As @caesarxuchao noted, it doesn't impact correctness and it can be refined and extended as time goes on.

As @liggitt has noted in the past, the leaking the storage version isn't required by clients and doesn't mean anything to every other client out there.

Choosing to trigger on this now doesn't preclude triggering on other items later on.  For a beta, this seems like a very simple start that doesn't require exposing extraneous information, doesn't limit future choice, and doesn't impact correctness.

lavalamp replied:
I view the kubernetes version and the api machinery version as orthogonal things. The apiserver binary version isn't logically correlated to the api versions. I would like to live in a world where I can keep my kubernetes version constant while picking up new api machinery changes, or vice versa.

Today these things happen to be correlated but that's accidental. It's really bad to make an automatic trigger that adds an unintended semantic to some facet of our api. E.g. is the /version endpoint even a "real" api covered by our deprecation policy? Adding this trigger changes the meaning of a version change from "I'm now running X" to "please touch every object in the system". I don't think that's a good meaning change. You should have to do something more than push a new patch to get every object in your system touched.

Just to be extra clear, my objection has nothing to do with efficiency and everything to do with not adding unintended consequences. 

I don't care about the exact mechanism we use as long as it has no false positives and no false negatives, again not because of efficiency but because that indicates it's being driven from a signal that's not going to change meaning in the future. We don't HAVE to expose the storage version, although that'd be a convenient thing to trigger on--it works great for rolling backwards as well as rolling forwards. Other clients don't need to know the storage version, but I think the storage migration triggerer component does deserve to know.

Are you worried that clients will abuse knowing the storage version? How?

I also really want some way of tracing resources when they move between groups, e.g. deployment. We haven't done too much of that recently, but I bet CRD authors are going to want to do it sooner or later.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,360 @@
+
+This is what the [Active Management][] design proposes. It requires:
+* exposing the configured [storage versions][] of Kubernetes built-in resource
+  types and aggregated resources. 
```
I still think this is the way to go. If an aggregated apiserver doesn't want to participate in the storage migrator or doesn't actually store its data or doesn't support ordinary changes or whatever, it can just omit the information in the discovery API.

deads2k commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,360 @@
+  // avoid conversions in the apiserver.
+  GroupVersions []GroupVersion
+  // Generation is the generation of the spec. Incrementing generation triggers
+  // migration. We do not use the objectMeta.generation because its semantic is
```
Pretty sure we don't bump it for that.  If we do, we should flag it as not good and avoid doing that.

caesarxuchao replied:
Yeah, for CR only spec change bumps the generation ([code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L118-L126)). I'll update.

The issue I cited was about CRD actually. And FWIW, deployment increments generation for annotations changes, see [code](https://github.com/kubernetes/kubernetes/blob/58d43abc39178dc2efc3b1ed6e4727bb34a4b68a/pkg/registry/apps/deployment/strategy.go#L106).

caesarxuchao replied:
Updated the KEP to use the `metadata.generation`.

deads2k commented:
> I don't care about the exact mechanism we use as long as it has no false positives and no false negatives, again not because of efficiency but because that indicates it's being driven from a signal that's not going to change meaning in the future.

Perhaps this is the conflict.  I don't care about an infrequent false positive.  A false positive just runs a rate limited read write cycle on resources.  

I think we can and should develop an API that can be triggered separately and I think the one described in this KEP is worth starting with.  If we separate the trigger mechanism out of the KEP for now, we can proceed on the guts behind them while we come to an agreement about how to trigger them.  In the simplest case, we simply provide a few simple trigger options and other integrators can share or build their own.

caesarxuchao commented:
Thanks for the input. I removed the contentious discussion on the the migration automation. PTAL.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+
+## Summary
+
+We propose to extend and improve the [oc adm migrate storage][] command line
```
missing push?

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,346 @@
+
+### API
+
+We introduce the `migration` API to record the intention and progress of a
```
It'd be OK to have a counter for the number of items migrated so far; that lets you see whether it's stuck or not which is pretty handy.

caesarxuchao replied:
Added `migration.status.continueToken`. User can use the token to check progress.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
+  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
+  // Immuntable.
```
Immutable

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+  // Immuntable.
+  Resource string 
+
+  // GroupVersions are all the groups and versions the apiserver serves this
```
Why is it necessary to pass all? Shouldn't a single one be sufficient? And why do you need to care whether it is preferred or not?

caesarxuchao replied:
> Why is it necessary to pass all?

For optimization. The migrator can skip the resource if this resource only has one version.

In the iteration addressing your comments, I updated the KEP to say that at the alpha stage, the tool migrates a hard-coded list of resources, this field became unnecessary.

> And why do you need to care whether it is preferred or not?

I thought if the migrator GET&PUT at the preferred version could save the apiserver a few conversions, but my calculation was wrong.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+  // avoid conversions in the apiserver.
+  GroupVersions []GroupVersion
+  
+  // LastUpdateTimestamp is the last time this spec is updated. The value is set
```
1. Why do you need to do this?
2. Please at least use a logical clock (an integer that counts up).

caesarxuchao replied:
Removed the generation stuff, so this is removed as well.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+  // which leads to the increment of `metadata.generation`.
+  LastUpdateTimestamp metav1.Time
+
+  // Reason of the migration. Examples are "manually triggered", "apiserver
```
I'm somewhat skeptical that this will be useful.

caesarxuchao replied:
Removed.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+// Note that the apiserver will increment `metadata.generation` when this spec
+// is updated.
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
```
I don't understand the relationship between Resource and GroupVersions.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,297 @@
+  // Reason of the migration. Examples are "manually triggered", "apiserver
+  // updated", etc.
+  Reason string
+}
```
I guess what I expected to see is an object that expresses a migration request for a single resource type (e.g., apps/v1/deployment). With status indicating progress {not started, migrated N objects, completed}, and where the object is removed after completion.

The generation stuff makes me think you want to reuse a single object to make multiple requests, but I don't understand why you would want to do that.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
````diff
@@ -81,158 +81,109 @@ As to the deliverables,
 ```golang
 type StorageVersionMigration struct {
   metav1.TypeMeta
+  // metadata.name should be "<resource>.<group>" of the resource being
````
Why? I'd probably just say to use generateName to avoid collisions. Requiring this means the old request has to be deleted before I can make a new request.

caesarxuchao replied:
I guess you want to keep the old requests to record history?

I'll have to add a `.status.active bool` to the `migration`, to let clients tell if the `migration` is for history keeping or is active. Sounds good?

lavalamp replied:
No, you already have a .status.completed for that. It's not about keeping history (which I don't want to do) it's about there not being any need to (effectively) have a per-resource lock on creating these things. (which is not even that effective since occasionally we move things between groups...)

Also imagine I have foo in group/v1 and also foo in group/v5, but they don't share key spaces because the one in v5 is a complete rewrite. Now I want to migrate v1 to v2 and also v5 to v6. It's contrived but I don't see any benefit at all from requiring the name match that format.

caesarxuchao replied:
Since kubernetes doesn't support transactions, I think it's easiest to only keep one `migration` for each resource type. We can keep the historic requests in a subfield.

caesarxuchao replied:
If the format of the name is `hash(etcd instance, etcd path prefix)` (as the next line in the comment suggested for the future), there is no issue in the case where we move things between groups, or in the case where group/v1/foo and group/v5/foo don't share key spaces.

It doesn't make sense to run multiple migration for the same resource type at the same time. If we go with your suggestion, for each resource type, we need to mark only one of the `migration` as active. Because Kubernetes doesn't support transactions, I don't see any easy way to do that. So I still favor the current design for simplicity.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -81,158 +81,109 @@ As to the deliverables,
+  Completed bool
+
+  // ContinueToken is the token to use in the list options to get the next chunk
+  // of objects to migrate. User can use this token to check the progress of the
```
Can the user use this to tell if the migration has started or not? If yes, how? If no, can we add this to the API?

caesarxuchao replied:
It would be empty if the migration hasn't started. And if I added `.status.active` to `migration` as in https://github.com/kubernetes/community/pull/2524#discussion_r221792276, that would indicate the start as well.

lavalamp replied:
Don't add a bunch of bools if an enum will serve.

Also for this one respond in the comment (responding on github is optional) ;)

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -81,158 +81,109 @@ As to the deliverables,
+  // ContinueToken is the token to use in the list options to get the next chunk
+  // of objects to migrate. User can use this token to check the progress of the
+  // migration.
+  ContinueToken string
```
Should this be optional?

caesarxuchao replied:
Sure. Added.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,249 @@
+
+### Alpha workflow
+
+We will create a **kube-migrator** command line tool. At the alpha stage,
```
I would greatly prefer CRDs or yamls that can be controlled by kubectl. Why would we add a new command line tool for this? The whole point of our control plane model is that it's extensible.

deads2k replied:
> I would greatly prefer CRDs or yamls that can be controlled by kubectl. Why would we add a new command line tool for this? The whole point of our control plane model is that it's extensible.

This is the basic tool we build up.  This tool would be what is later trigger automatically.  Doing it like this allows us to build up a series of useful pieces.  Nothing precludes adding controller logic around it (this doc describes exactly that), but a standalone CLI tool has value in and of itself.

caesarxuchao replied:
I think Daniel also supports building up the migrator in separable pieces. His concrete [suggestion] (https://github.com/kubernetes/community/pull/2524#discussion_r221402698) was moving the API discovery step to a separate job, that would leave very little logic in the CLI.

I think we should still make the CLI, it will be a thin layer that launches jobs and creates CRs. Users should be able do all the operations via kubectl. The value of the CLI is saving users from editing yamls.

caesarxuchao replied:
Reading the KEP again, I think both the "kube-migrator install" and "kube-migrator progress" commands can be replaced with kubectl one-liners. So I removed the standalone CLI.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -81,158 +81,109 @@ As to the deliverables,
 * deletes any existing `deployment` of **kube-migrator controller** 
 * creates a `deployment` of **kube-migrator controller** 
+* generates a comprehensive list of resource types via the discovery API
+* discovers all custom resources via listing CRDs
```
We can't tell if it's a CRD or a built-in resource by looking at the apiservices. The apiservices for CRD also have `.spec.service == null`.

lavalamp replied:
...which doesn't matter since you want to exclude both, no?

caesarxuchao replied:
I want to exclude both the custom resources (CR) and the aggregated resources (AR). I want to only include the built-in resources. From the apiservices, I can tell AR from CR and the built-ins, but I can't tell CR from the built-ins.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -81,158 +81,109 @@ As to the deliverables,
+* removes the custom resources and aggregated resources from the comprehensive
+  resource list. The list now only contains Kubernetes built-in resources.
+* removes resources that share the same storage. At the alpha stage, the
+  information is hard-coded, like in this [list][].
```
The link works for me. Are you referring to the empty `[]`? That's a neat format of inserting hyperlinks in git markdown. It has a pairing `[list]:https://github.com/openshift/origin/blob/2a8633598ef0dcfa4589d1e9e944447373ac00d7/pkg/oc/cli/admin/migrate/storage/storage.go#L120-L184` at the end of the section. If you `view` the rendered file you'll be able to click through.

lavalamp replied:
ah my bad.

deads2k commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,249 @@
+
+type StorageVersionMigrationStatus {
+  // Completed means the migration of this resource has completed.
+  Completed bool
```
I expected status to include conditions which are fairly standard in the API and have built in scripting functionality.

caesarxuchao replied:
Do you mean making it a enum supporting states like Running, Succeeded, Failed? I agree it should be an enum instead of an boolean.

I'm not sure what you mean by "built in scripting functionality"?

caesarxuchao replied:
I'll start with adding `.status.phase`.

caesarxuchao replied:
Ok, I've converted to conditions.

deads2k commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,249 @@
+type StorageVersionMigrationSpec {
+  // Resource is the resource that is being migrated. The migrator sends
+  // requests to the endpoint serving the Resource.
+  // Immutable.
```
if this is immutable, how do I update a field in here to indicate that the generation needs to be bumped?  The reason field made such a thing a possible and allowed me to key a request to an external system with an opaque key.

caesarxuchao replied:
I'm thinking in the future we will add a mutable field to this spec, like `targetStorageVersion(Hash)`, then the spec will be changing.

My concern with a `Reason string` field is it's too loose. User might edit the `Reason` and inadvertently triggered a new round of migration.

Can the external system to be triggered by the `migration.status.conditions == proposed`?

deads2k replied:
> Can the external system to be triggered by the `migration.status.conditions == proposed`?

The external system drives this migration.  It needs the generation of this object to be incremented to do that.

lavalamp replied:
Why does the generation need to change if you are deleting and recreating to request a new migration?

caesarxuchao replied:
I think three of us had three ideas on how to drive the migration control loop. I updated the KEP to follow Daniel's idea.

The original three ideas are:

I wanted to maintain one `migration` CR for one resource type, and the control loop would start the migration if `migration.status==nil`. User needs to delete and create `migration` to request for another round of migration.

I think deads2k's model was also one `migration` CR for one resource type, but using the `migration.metadata.generation != observedGeneration` to drive the control loop to do migration.

In lavalamp's model, users *can* create multiple `migration` CRs for a single resource type. Each CR will cause the control loop to do the migration once. This results in wasteful migration, but doesn't affect correctness. This state is recoverable by deleting redundant `migration`s. And we won't reach this state unless users are manually creating duplicate `migration` CRs. I updated the KEP to lavalamp's idea because it's the simplest and low risk.

deads2k replied:
> I think three of us had three ideas on how to drive the migration control loop. I updated the KEP to follow Daniel's idea.
> 
> The original three ideas are:
> 
> I wanted to maintain one `migration` CR for one resource type, and the control loop would start the migration if `migration.status==nil`. User needs to delete and create `migration` to request for another round of migration.
> 
> I think deads2k's model was also one `migration` CR for one resource type, but using the `migration.metadata.generation != observedGeneration` to drive the control loop to do migration.
> 
> In lavalamp's model, users _can_ create multiple `migration` CRs for a single resource type. Each CR will cause the control loop to do the migration once. This results in wasteful migration, but doesn't affect correctness. This state is recoverable by deleting redundant `migration`s. And we won't reach this state unless users are manually creating duplicate `migration` CRs. I updated the KEP to lavalamp's idea because it's the simplest and low risk.

I think that I could use the model multiple migration CRs for a single resource type.  It's a more batch-like way of looking at the problem, but I think it can work.  Proof of the pudding is in the tasting - I'm willing to taste it and see.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -81,158 +81,109 @@ As to the deliverables,
+* creates `migration` CRs (see the [API section][] for the schema) for all
+  remaining resources. The `ownerReferences` of the `migration` objects are set
+  to the **kube-migrator controller** `deployment`. Thus, the old `migration`s
+  are deleted with the old `deployment` in the first step.
```
I don't follow. I'm think that the user needs to recreate the initialization `job`, maybe by `kubectl apply` the `job` with a different name. The `job` will delete the old controller and `migration`s and create new ones.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
-  // metadata.name can be hash(etcd instance, etcd path prefix).
+  // At the alpha stage, metadata.name must be "<resource>.<group>" of the
+  // resource being migrated.
+  // We require this name format to avoid duplicate `migration` CRs for the same
```
I still think preventing such a duplicate should be a non-goal; it prevents you from starting a new one if a record of an old one is still around.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
+  // resource being migrated.
+  // We require this name format to avoid duplicate `migration` CRs for the same
+  // resource type. Otherwise we will need to build complicated system to make
+  // sure the latest `migration` CR is honored. 
```
You already execute only one migration at a time; why would you need to build anything else?

lavalamp replied:
Why is that problematic? Do you expect clients to request that? Folks have repeatedly said we don't care about extra migrations.

Also if the client wants to do that, they can always just repeatedly delete and re-add the object.

We don't need to protect them from themselves to that degree.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
+  // b. groupX/v1/foo and groupX/v5/foo may not share key spaces and thus
+  // require two `migration` CRs, though Kubernetes don't have such cases yet. 
+  // To address the two concerns, in the beta version, the name need to be
+  // tied to the key spaces, e.g., a hash of "etcd instance"+"etcd path prefix".
```
The only reason I can think of for layering requirements on the name is for idempotency. Is that why you are doing it?

Why can't you just let the client worry about that? Why do you have to mandate it?

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
   metav1.ObjectMeta
   Spec StorageVersionMigrationSpec
+  // If status is empty, it means the `migration` is just created. The migrator
+  // controller hasn't launched the **migration worker** `job` for it yet.
```
All of our objects work like this, is it necessary to state?

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
+type MigrationPhase string
+
+const (
+  // MigrationPending indicates that the migration is requested, but no 
```
This contradicts what you said about "status"?

caesarxuchao replied:
Right. I forgot to remove it.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
+  // MigrationSucceed indicates that the migration has completed successfully.
+  MigrationSucceeded MigrationPhase = "Succeeded"
+  // MigrationSucceed indicates that the migration has failed.
+  MigrationRunning MigrationPhase = "Running"
```
Comment, name, and value are copied wrong :)

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
-  // Completed means the migration of this resource has completed.
-  Completed bool
+  // Phase indicates where the migration is in the life-cycle. Do not treat it
+  // as a state machine, because new phases can be added.
```
They can?

I think this is an impossible ask from clients.

There is confusion about this, but I think the latest guidance is still that extensible things like this should be represented as conditions.

But does this need to be extensible?

caesarxuchao replied:
Converted to the conditions pattern.

lavalamp commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
-  // migration.
+  // of objects to migrate. When the .status.phase is "Running", users can use
+  // this token to check the progress of the migration.
+  // +optional
```
To be extremely technical, continue token isn't status, because it's not reproducible from the environment.

caesarxuchao replied:
The migrator can maintain a configmap to record the continue token. That would make this field reproducible, but it doesn't add any value.

caesarxuchao replied:
Done.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
````diff
@@ -0,0 +1,277 @@
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
````
Added a line saying that we abbreviated it as `migration` in the text.

I think the API name should be fully qualified, because we might add other kind of migrations in the future.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
````diff
@@ -0,0 +1,277 @@
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  // metadata.generateName will be "<resource>.<group>" of the resource being
````
It's only intended for the readers of this KEP. I'll remove it in the API PR. Explained in the KEP.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,277 @@
+
+## Proposal
+
+### Alpha workflow
```
Added in the workflow section.

caesarxuchao commented on keps/sig-api-machinery/storage-migration.md:
```diff
@@ -0,0 +1,277 @@
+  Resource GroupVersionResource
+}
+
+type MigrationConditionType string
```
We might add other conditions like "Pause"? There will be other clients monitoring the `migrations.status`, for example, cloud providers monitor this status to decide if the cluster is safe to upgrade. `conditions` force developers to not think the status as a state machine.

Also the `kubectl wait` function only works with `.status.conditions`. The command is marked as "experimental" but it's quite neat.

lavalamp commented:
/approve

perhaps you should squash?

@deads2k it's changed a bit since you last saw it, do you want to look again?

lavalamp commented:
Looks like owners has changed.

/assign @idvoretskyi

calebamiles commented:
/approve
